### DISCUSSION TITLE
How to pass time filters to SQL Lab queries

### DISCUSSION THREAD

**Discussion by cachafla** [apache/superset#8183] [General]
Is there a way to pass `[from, to]` date filters to an SQL query?

I've been looking everywhere in documentation and source code and haven't found a solution to passing date filters to custom SQL queries. I'm working on some advanced multi-join aggregations from existing tables/views.

As a very simple example of the issue I'm trying solve, consider this SQL query:

```sql
SELECT COUNT(*)
FROM sessions
WHERE timestamp <= {{ something }} AND timestamp >= {{ something_else }}
```

I'd like to be able to access the date filters that are propagated through a `Filter Box` widget. in a dashboard. I've found that the values of these filters are available as these variables (thanks to https://superset.incubator.apache.org/faq.html#how-to-add-dynamic-filters-to-a-dashboard): https://github.com/apache/incubator-superset/blob/40776bd547166b6975574e815ec02add04781248/superset/assets/src/visualizations/FilterBox/FilterBox.jsx#L38-L44

With that in mind, I could make use of `__time_range` inside my SQL filter via `filter_values('__time_range')`, but `__time_range` is always giving me the literal time range selected, not a parsed `[from, to]` couple of date objects. I get values such as:

- 'Last quarter'
- '2019-09-02T00:00:00 : 2019-09-03T00:00:00'

These variations are very difficult to deal with inside an SQL query.

I'd like to know if this is all I get or if there's an undocumented solution to my problem I don't know about. In its simplest from, I'd like to have filters available in the SQL as template params so that I can do things like:

```sql
timestamp >= {{ from }} AND timestamp <= {{ to }}
```

**Comment by lilila** [comment]
Is there any update on this topic? We also tried to get the values of time range filter using `filter_values` but with superset 1.0.rc4 we get no value (not even a string value like the one mentioned by @cachafla )

The problem we are facing is the following: 
We have time series and we would like to plot them aligned (independently of the date). To do so, we compute a fake_date based on the difference in days between the first available date and the current date. Thus all figures for Day1, Day2... are aligned for the different lines. It works well but when we want to add a time filter, then it is applied to the fake date and it breaks the logic. 

It would be nicer if : 
- aligning time series based on the first available observation is an option of the plot. 
- or to allow the line chart to take as x_axis integer (like day num) instead of date
- or the best would be to allow the user to get the information from the time range filter (as other filters values), in the case described above it would allow us to select data from the right period of time and then create dynamically a fake_date that would fit the time range selection.

More generally, why is the time range filter different from the other filters?

**Reply by caiobelfort** [reply]
I think the variables {{ from_dttm }} and {{ to_dttm }} can be used to retrieve a date range filter values.

**Reply by k7daniel** [reply]
But these values are not available while in SQL Lab or when verifying a datasets columns?

**Reply by kevin-brendle** [reply]
> But these values are not available while in SQL Lab or when verifying a datasets columns?

in SQL Lab you just need to define them with 
```
{% set from_dttm = '2022-01-11 23:00:00' %}
{% set to_dttm = '2022-01-11 23:10:00' %}
```
and then your query will work

**Comment by TheRum** [comment]
I am also facing the same issue in version 1.0.0 . It was working in prior versions but is broken now.

Here's the function I have defined in superset_config.py .


```
from superset.utils.date_parser import get_since_until
from typing import Optional, Any
from flask import request
import json

def time_filter(default: Optional[str] = None) -> Optional[Any]:
    form_data = request.form.get("form_data")

    if isinstance(form_data, str):
        form_data = json.loads(form_data)
        extra_filters = form_data.get("extra_filters") or {}
        time_range = [f["val"] for f in extra_filters if f["col"] == "__time_range"]
        time_range = time_range[0] if time_range else None

        since, until = get_since_until(time_range)
        time_format = '%Y-%m-%d %H:%M:%S'

        until = until.strftime(time_format)
        if not since:
            return '<= \'{}\''.format(until)
        since = since.strftime(time_format)
        return 'BETWEEN \'{}\' AND \'{}\''.format(since, until)
    return default

JINJA_CONTEXT_ADDONS = {
    'time_filter': time_filter
}
```

**Reply by lilila** [reply]
thank you for sharing this

**Reply by elaragon** [reply]
Yes! I was unable to upgrade superset because the previous jinja function was no longer working in new versions, thanks :)

**Reply by GiannisDimitriou** [reply]
@TheRum I think Jinja functionality was put behind a flag and it was disabled by default few releases before.
You can check the changelog if you didn't do it already

**Reply by TheRum** [reply]
@GiannisDimitriou Thanks for the suggestion. It really helped me!
I think we should improve superset documentation to include this information. This is really helpful and people like me might have wasted a lot of time behind this.

Here's what I have tried and working for me !

> superset_config.py
```
DEFAULT_FEATURE_FLAGS = {
    "ENABLE_TEMPLATE_PROCESSING": True
}
```

> SQL
```
select *
from druid.xyz
where 1 = 1
{% if from_dttm %} and __time >= TIMESTAMP '{{ from_dttm }}'{% endif %}
{% if to_dttm %} and __time < TIMESTAMP '{{ to_dttm }}'{% endif %}
{% set state = filter_values("State") %}
{% if state %} AND State IN ({{ "'" + "', '".join(state) + "'" }}){% endif %}
```


Also, If you want to have your own function to be implemented, here's the one !

> superset_config.py
```
from typing import Optional
from superset.utils.date_parser import get_since_until
from superset.utils.core import merge_extra_filters

def time_filter() -> Optional[str]:
    from superset.views.utils import get_form_data

    form_data, _ = get_form_data()
    merge_extra_filters(form_data)

    time_range = form_data.get("time_range")

    since, until = get_since_until(time_range)
    time_format = '%Y-%m-%d %H:%M:%S'

    if not until:
        return None
    until = until.strftime(time_format)

    if not since:
        return '<= \'{}\''.format(until)
    since = since.strftime(time_format)

    return 'BETWEEN \'{}\' AND \'{}\''.format(since, until)

JINJA_CONTEXT_ADDONS = {
    'time_filter': time_filter
}
```

> SQL
```
{% set dttm = time_filter() %}
{% if dttm %} AND __time {{ dttm }}{% endif %}
```

**Reply by qlpanalytics** [reply]
Hi @TheRum! I was hoping you could provide some insight on an issue I'm having with this. I'm trying to reference the date range in a NESTED select statement within a metric. The parameters '{{from_dttm}}' and '{{to_dttm}}' work in the virtual SQL dataset, but not within the nested select statement metric. Do you have any ideas why?

Additional information can be found in my question here:
https://github.com/apache/superset/discussions/16117

Any suggestions are appreciated!

**Reply by TheRum** [reply]
@qlpanalytics It seems jinja template is rendered only for the virtual sql datasets and not for the metric sql statements.
Let me have a look if I can find something on this.

**Reply by qlpanalytics** [reply]
Thanks for checking into it, @TheRum! We are stumped, so keeping our fingers crossed that you're able to find some sort of workaround/solution :)

**Reply by TheRum** [reply]
@qlpanalytics I checked in the documentation and issues but nowhere it is mentioned whether it is supported or not. So, I believe templating is not supported in the calculated columns/metrics. I tried as well but no luck.
I think you should ask Preset team via email or slack, maybe they can confirm if it is supported or not. And if yes, how you can implement the same.

Hope this helps 🙂.

**Reply by qlpanalytics** [reply]
@TheRum thank you so much for the follow up! I will reach out to the team as suggested. If this turns out to be possible, I will update this thread with the solution. Thanks again!

**Reply by MaximSlagter** [reply]
@qlpanalytics Any updates? ;)

**Reply by qlpanalytics** [reply]
@MaximSlagter unfortunately this did not work and we were unable to complete the project as intended in Superset. We were successful in achieving this in Google Data Studio though - better than nothing :)
